torch/_masked/__init__.py

```diff
@@ -443,15 +499,48 @@ def sum(input: Tensor,
     # __doc__ is generated by _apply_docstring_templates decorator
     if dtype is None:
         dtype = input.dtype
-    # TODO: What follows is a reference implementation of a masked sum
-    # operation that is to be replaced with an optimized one and
-    # extended to support other layouts.
+    dim_ = _canonical_dim(dim, input.ndim)
+
     if input.layout == torch.strided:
+        mask = _input_mask(input, mask=mask)
         mask_input = input if mask is None else torch.where(mask, input, input.new_zeros([]))
-        dim_ = _canonical_dim(dim, input.ndim)
         return torch.sum(mask_input, dim_, bool(keepdim), dtype=dtype)
+
+    elif input.layout == torch.sparse_coo:
+        input = input.coalesce()
+        if mask is None or mask is input:
+            mask_input = input
```

Comment:
If mask is input, _input_mask could convert it to a mask via a call to `to.(torch.bool)` or we require the user to always provide a bool Mask (which I think is a good idea to keep the API simpler).

Reply:
Resolving via https://github.com/pytorch/pytorch/pull/71239#discussion_r831303602